lib/src/lazy_load_scrollview.dart

```diff
@@ -1,89 +1,296 @@
+  /// The debounce strategy is eager. That means the callback will be called
+  /// first then further calls will be ignored until the timeout expires.
+  final Duration debounceDuration;
+
+  /// Whether to ignore the depth of the [ListView] when calling [onLoadBefore]
+  /// or [onLoadAfter]. This value is ignored if [controller] is set.
+  final bool ignoreDepth;
+
+  /// A check that specifies whether a [ScrollNotification] should be
+  /// handled by this widget.
+  ///
+  /// By default, checks whether `notification.depth == 0`. That means if the
+  /// scrollbar is wrapped around multiple [ScrollView]s, it only responds to the
+  /// nearest scrollView and shows the corresponding scrollbar thumb.
+  ///
+  /// Copied from [RawScrollBar]
+  ///
+  /// This value is ignored [controller] is non null or [ignoreDepth] is true.
+  final ScrollNotificationPredicate? notificationPredicate;
 
-  LazyLoadScrollView({
+  /// This widget can either listen to notifications through a
+  /// [NotificationListener] or it can listen to a [ScrollListener] and handle
+  /// the scroll events directly.
+  final ScrollController? controller;
+
+  const LazyLoadScrollView({
     Key? key,
-    required this.child,
-    required this.onEndOfPage,
-    this.scrollDirection = Axis.vertical,
-    this.isLoading = false,
-    this.scrollOffset = 100,
+    required this.builder,
+    this.onLoadAfter,
+    this.onLoadBefore,
+    this.scrollOffset = 100.0,
+    this.debounceDuration = const Duration(milliseconds: 500),
+    this.timeout = const Duration(seconds: 10),
+    this.ignoreDepth = false,
+    this.notificationPredicate = defaultScrollNotificationPredicate,
+    this.controller,
   }) : super(key: key);
+
+  @override
+  State<LazyLoadScrollView> createState() => _LazyListBuilderState();
 }
 
-class LazyLoadScrollViewState extends State<LazyLoadScrollView> {
-  LoadingStatus loadMoreStatus = LoadingStatus.STABLE;
+class _LazyListBuilderState extends State<LazyLoadScrollView> {
+  AxisDirection? scrollDirection;
+
+  bool isLoadingBefore = false;
+
+  bool isLoadingAfter = false;
+
+  Timer? timer;
+
+  Future<void> loadBefore() async {
+    if (widget.onLoadBefore == null || isLoadingBefore) return;
+
+    // If we have a timer going, just return early.
+    if (timer?.isActive ?? false) {
+      return;
+    } else {
+      setState(() => isLoadingBefore = true);
+
+      if (mounted) {
+        await widget.onLoadBefore?.call().timeout(widget.timeout);
+        setState(() => isLoadingBefore = false);
+      }
+
+      timer = Timer(widget.debounceDuration, () {});
+    }
+  }
+
+  Future<void> loadAfter() async {
+    if (widget.onLoadAfter == null || isLoadingAfter) return;
+
+    if (timer?.isActive ?? false) {
+      return;
+    } else {
+      setState(() => isLoadingAfter = true);
+
+      if (mounted) {
+        // TODO: Will these need completers to correctly handle errors?
```

Comment:
What would you expect to do with the error?

Reply:
It's user convenience. I have a package called async_button_builder that wraps a user's callback and handles its various states. I found that if I don't use a completer and their function errors, I can `rethrow` but that won't give them the stackTrace. It'll also throw at my callsite of their function too rather than in their original function which was kinda lame. It's a gut assumption right now as I haven't tested, but I think if their load more functions error, the same will happen and they'll get no helpful stacktrace (again, not sure tho). I'll test on weekend

Reply:
Ah, nvm, just checked real quick. It handles the error correctly. I'll remove that TODO:

Reply:
I did forget though that if it errors, it won't call anything afterward which means the loading state won't be reset. Here's before:

```dart
await widget.onLoadAfter?.call().timeout(widget.timeout);
setState(() => isLoadingAfter = false);
```

after:
```dart
await widget.onLoadAfter?.call().timeout(widget.timeout).whenComplete(
  () {
    setState(() => isLoadingAfter = false);
  },
);
```

Reply:
6e41d019b3801c09554b311fbcce992abc9016d9